src/main/java/org/radarcns/stream/phone/PlayStoreLookup.java

```diff
@@ -0,0 +1,152 @@
+import org.jsoup.nodes.Element;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A Google Play Store lookup backed by a cache.
+ */
+public final class PlayStoreLookup {
+    private static final Logger log = LoggerFactory.getLogger(PlayStoreLookup.class);
+    private static final String URL_PLAY_STORE_APP_DETAILS = "https://play.google.com/store/apps/details?id=";
+    private static final String CATEGORY_ANCHOR_SELECTOR = "a.document-subtitle.category";
+
+    private final long cacheTimeoutSeconds;
+    // Do not cache more than 1 million elements, for memory consumption reasons
+    private final LRUCache<String, AppCategory> categoryCache;
+
+    public PlayStoreLookup(long cacheTimeoutSeconds, int maxCacheSize) {
+        this.cacheTimeoutSeconds = cacheTimeoutSeconds;
+        this.categoryCache = new LRUCache<>(maxCacheSize);
+    }
+
+
+    /**
+     * Looks up a category from the play store, if possible from cache.
+     *
+     * The app category may have a null category name:
+     * - if the page could not be retrieved (not public)
+     * - category element on play store is not available
+     * - no connection can be made with the play store
+     *
+     * In the last event, the category name is not cached, so it will be retried in a next call.
+     *
+     * @param packageName name of the package as registered in the play store
+     * @return category as given by the play store
+     */
+    public AppCategory lookupCategory(String packageName) {
+        // If not yet in cache, fetch category for this package
+        AppCategory category = categoryCache.get(packageName);
+
+        long cacheThreshold = System.currentTimeMillis() / 1000L - cacheTimeoutSeconds;
+        if (category == null || category.fetchTimeStamp < cacheThreshold) {
+            try {
+                category = fetchCategory(packageName);
+                categoryCache.put(packageName, category);
+            } catch (IOException ex) {
+                // do not cache anything: we might have better luck next time
+                log.warn("Could not find category of {}: {}", packageName, ex.toString());
+                category = new AppCategory(null);
+            }
+        }
+
+        return category;
+    }
+
+    /**
+     * Fetches the app category by parsing apps Play Store page.
+     *
+     * The app category may have a null category name:
+     * - if the page could not be retrieved (not public)
+     * - category element on play store is not available
+     *
+     * @param packageName name of the package as registered in the play store
+     * @return category as given by the play store
+     * @throws IOException if no connection can be made with the Google App Store
+     */
+    public static AppCategory fetchCategory(String packageName) throws IOException {
+        String url = URL_PLAY_STORE_APP_DETAILS + packageName;
+
+        try {
+            Document doc = Jsoup.connect(url).get();
```

Comment:
Are we sure this implementation works at scale? If `Google` applies some quotas or scraper checks, it may block or slow this call. If this is the case, we could mitigate it using a public [Http proxy](https://www.proxynova.com/proxy-server-list/) or checking if frequently changing the `user-agent` allow us to avoid the smart test.